src/circuit/gadgets/constraints.rs

```diff
@@ -486,4 +1006,240 @@ mod tests {
+            cond,
+            alloc_num,
+            alloc_most_positive,
+        );
+        assert!(res.is_ok());
+        assert!(cs.is_satisfied());
+    }
+
+    #[test]
+    fn test_enforce_less_than_bound() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+        let num = s.num(42);
+        let alloc_num =
+            AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "num"), s, || Ok(&num)).unwrap();
+        let bound = s.num(43);
+        let alloc_bound =
+            AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "bound"), s, || Ok(&bound)).unwrap();
+        let cond = Boolean::Constant(true);
+
+        let res = enforce_less_than_bound(
+            &mut cs.namespace(|| "enforce less than bound"),
+            cond,
+            alloc_num,
+            alloc_bound,
+        );
+        assert!(res.is_ok());
+        assert!(cs.is_satisfied());
+    }
+
+    #[test]
+    fn test_enforce_less_than_bound_negative() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+        let num = s.num(43);
+        let alloc_num =
+            AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "num"), s, || Ok(&num)).unwrap();
+        let bound = s.num(42);
+        let alloc_bound =
+            AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "bound"), s, || Ok(&bound)).unwrap();
+        let cond = Boolean::Constant(true);
+
+        let res = enforce_less_than_bound(
+            &mut cs.namespace(|| "enforce less than bound"),
+            cond,
+            alloc_num,
+            alloc_bound,
+        );
+        assert!(res.is_ok());
+        assert!(!cs.is_satisfied());
+    }
+
+    #[test]
+    fn test_enforce_n_bits() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+
+        let g = GlobalAllocations::new(&mut cs.namespace(|| "global_allocations"), s).unwrap();
+
+        let num = crate::Num::<Fr>::Scalar(Fr::from_u64(42).unwrap()); // 42 = 101010
+        let alloc_num =
+            AllocatedNum::alloc(&mut cs.namespace(|| "num"), || Ok(num.into_scalar())).unwrap();
+
+        let res = enforce_n_bits(&mut cs.namespace(|| "enforce n bits"), &g, alloc_num, 6);
+        assert!(res.is_ok());
+        assert!(cs.is_satisfied());
+    }
+
+    #[test]
+    fn test_enforce_n_bits_negative() {
+        let mut cs = TestConstraintSystem::<Fr>::new();
+        let s = &mut Store::<Fr>::default();
+
+        let g = GlobalAllocations::new(&mut cs.namespace(|| "global_allocations"), s).unwrap();
+
+        let num = crate::Num::<Fr>::Scalar(Fr::from_u64(42).unwrap()); // 42 = 101010
+        let alloc_num =
+            AllocatedNum::alloc(&mut cs.namespace(|| "num"), || Ok(num.into_scalar())).unwrap();
+
+        let res = enforce_n_bits(&mut cs.namespace(|| "enforce n bits"), &g, alloc_num, 5);
+        assert!(res.is_ok());
+        assert!(!cs.is_satisfied());
+    }
```

Comment:
This and the previous test (`test_enforce_n_bits`) seem to be showing that 42 is both 5 (this test) and 6 (the previous) bits. Am I misreading?

If not, it's unclear what the expected behavior of `enforce_n_bits` should be. Based on the name, I would expect that only one possible value would be correct. Considering only this issue, maybe the function means to be named `enforce_at_least_n_bits`?

Alternately, maybe there is an error in the function. If I've instead misread the tests, let me know. It might be that some documentation of the tests' intention would clarify that.

Reply:
The one called *_negative (using number 6) asserts the circuit is NOT satisfied.

Reply:
Aha, that explains it. Thanks.

Although I think you mean using number 5, right? This makes a difference per my comment on the function itself.

That is, I expect that the first test (with 6) shows that 42 is 'at most 6 bits', but the test asserting it is 'at most 5 bits' fails because it is in fact 6 bits (> 5).

Reply:
yes, 5

Reply:
We are enforcing the number is less than 2ˆn.

Reply:
Can we rename the function to one that doesn't seem to imply exactness?

Reply:
done